vmm/src/vstate.rs

```diff
@@ -280,6 +292,84 @@ pub struct Vcpu {
 }
 
 impl Vcpu {
+    thread_local!(static TLS_VCPU_PTR: VcpuCell = Cell::new(None));
+
+    /// Associates `self` with the current thread.
+    ///
+    /// Is a prerequisite to succesfully run `run_on_thread_local_vcpu()` on the current thread.
+    /// Will fail if there is already a `Vcpu` present in this thread's TLS.
+    fn init_thread_local_data(&mut self) -> Result<()> {
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if cell.get().is_some() {
+                return Err(Error::VcpuTlsInit);
+            }
+            cell.set(Some(self as *mut Vcpu));
+            Ok(())
+        })
+    }
+
+    /// Deassociates `self` from the current thread.
+    ///
+    /// Should be called if the current `self` had called `init_thread_local_data()` and
+    /// now needs to move to a different thread.
+    ///
+    /// Fails if `self` was not previously associated with the current thread.
+    fn reset_thread_local_data(&mut self) -> Result<()> {
+        // Best-effort to clean up TLS. If the `Vcpu` was moved to another thread
+        // _before_ running this, then there is nothing we can do.
+        let mut tls_contains_self = false;
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if let Some(vcpu_ptr) = cell.get() {
+                if vcpu_ptr == self as *mut Vcpu {
+                    tls_contains_self = true;
+                }
+            }
+        });
+
+        if tls_contains_self {
+            Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| cell.take());
+            Ok(())
+        } else {
+            Err(Error::VcpuTlsNotPresent)
+        }
+    }
+
+    /// Runs `func` for the `Vcpu` associated with the current thread.
+    ///
+    /// It requires that `init_thread_local_data()` was run on this thread.
+    ///
+    /// Fails if there is no `Vcpu` associated with the current thread.
+    fn run_on_thread_local_vcpu<F>(func: F) -> Result<()>
+    where
+        F: FnOnce(&mut Vcpu),
+    {
+        Self::TLS_VCPU_PTR.with(|cell: &VcpuCell| {
+            if let Some(vcpu_ptr) = cell.get() {
+                let vcpu_ref: &mut Vcpu = unsafe { &mut *vcpu_ptr };
+                func(vcpu_ref);
+                Ok(())
+            } else {
+                Err(Error::VcpuTlsNotPresent)
+            }
+        })
+    }
+
+    /// Registers a signal handler which makes use of TLS and kvm immediate exit to
+    /// kick the vcpu running on the current thread, if there is one.
+    pub fn register_vcpu_kick_signal_handler() {
+        unsafe {
+            extern "C" fn handle_signal(_: c_int, _: *mut siginfo_t, _: *mut c_void) {
+                let _ = Vcpu::run_on_thread_local_vcpu(|vcpu| {
```

Comment:
Do we ever expect an error here during normal operation? Right now it seems we're ignoring the result, but if an error only appears due to a severe program logic problem on our part, should we consider terminating the process here? What can happen if `run_on_thread_local_vcpu` returns an error and we continue running?

Reply:
In general, terminating the process because of failing to kick the vcpu is too extreme. It is definitely a recoverable error. The microVM can continue normal functionality with no problem in such a case, I wouldn't recommend violently terminating it.

In this particular case, the only error case is the one where the TLS was not initialized and there is no `Vcpu` present. This failure will be later caught in the communication mechanism between the the `Vcpu` thread and its controlling thread when the `Vcpu` will fail to respond to messages (as it has not been successfully kicked).